### ISSUE TITLE
Horovod setup for 4 individual machine with 1 GPU each

### ISSUE THREAD

**Issue by niranjansuthar70** [horovod/horovod#2553]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Hello,
I am planning to use Horovod for distribution training with host machine (1 GPU), and other 3 machine with 1 GPU each.
So should I install Horovod on each machine? and how can I create these server1,2,3,4 mention here? "$ horovodrun -np 4 -H server1:1,server2:1,server3:1,server4:1 python train.py"

Also how can I pin these 4 GPUs in these lines:

"config = tf.ConfigProto()
config.gpu_options.visible_device_list = str(hvd.local_rank())"

thanks!

**Comment by tgaddair** [comment]
Hey @niranjansuthar70, to answer your questions:

1. Yes, you should install Horovod in each machine.
2. `server1` etc. are the IP addresses or hostnames of the machines in your network.  You can create them using Kubernetes, EC2, etc. or run directly on machines in a private cluster.
3. Since each machine has one GPU, you just need to make sure that you have on process running per GPU, then the local rank will always be 0, so each process will use the 1 GPU available.

Let me know if you have additional questions.

**Comment by niranjansuthar70** [comment]
@tgaddair thanks for your response.
1. should I run my train.py in all the 4 machines (with 1 GPU each)? ["$ horovodrun -np 4 -H server1:1,server2:1,server3:1,server4:1 python train.py"]
2. And also do I need to copy dataset in all machines?

**Comment by chongxiaoc** [comment]
> @tgaddair thanks for your response.
> 
> 1. should I run my train.py in all the 4 machines (with 1 GPU each)? ["$ horovodrun -np 4 -H server1:1,server2:1,server3:1,server4:1 python train.py"]
> 2. And also do I need to copy dataset in all machines?

1 .yes
2. This is same as your question of `install Horovod on each machine?`
Backend of horovod is mpi. MPI is `SPMD` model which requires filesystem on all nodes are synced.
It is same for installing `horovod`, installing `mpi` as well as `putting` data.

**Comment by arpit-debug** [comment]
@tgaddair we are able to setup the horovod on the 2 system 
there are two approach we are updating the gradient  async parameter server and all sync training 
my questions:
1. how to specify in the code that which approach should use by horovod? (where the modification required in the code to change the approach)
and in your opinion
2. which approach should be better async parameter server or all sync?

**Comment by tgaddair** [comment]
Hey @arpit-debug, Horovod only supports synchronous SGD.  We do not support async parameter server based training, as in most cases our users have had model convergence issues using async SGD.

**Comment by arpit-debug** [comment]
Thank you for reply @tgaddair i still don't have clarification on the some of the points can you provide the some insights on these points
1. Rate of convergence of the loss are almost same with multiple distribution and single system. so it is expected or not? if loss convergence rate is same with multiple distribution and single system then where we  are getting advantage?
2. With Respect to loss convergence which option will be better? providing the whole data on all the system or splitting the data on all system?

**Comment by tgaddair** [comment]
Hey @arpit-debug, with distributed training, you should see loss converge at the same rate per epoch, but each epoch should be significantly faster.  That's because you're processing multiple batches in parallel on different workers.  In order for this to work, you need to split the data across the workers (data parallelism).

**Comment by arpit-debug** [comment]
Thanks for quick response @tgaddair 
1.we have to split the data manually on every worker or horovod gonna split it on own?
2. if data is distributed on the all the workers how horovod counting the 1 epoch? like all the images completed on every worker that's one epoch ?

**Comment by tgaddair** [comment]
Hey @arpit-debug, Horovod does not split the data itself, you could use a framework like [Petastorm](https://github.com/uber/petastorm) to do this, or have each worker read a specific chunk of the data manually.

Horovod counts epochs by a full pass over all the data.  For example, if you have 100 examples in your dataset, batch size 1, and 1 worker, then it will take 100 steps to process one epoch.  If you instead had 4 Horovod workers each with batch size 1, it will take 25 steps to process 1 epoch (because you've divided the data among the workers).  Does that make sense?

**Comment by niranjansuthar70** [comment]
Hello @tgaddair , I have successfully built horovod in miniconda env and I have checked nccl,mpi,tf in "`horovodrun --check-build`".
I have 2 machines with same minoconda env along with same horovod packages. Now should I run my main script on both machine env at same time? I have enabled ssh passwordless on both machines. How should I tell horovodrun to run same script on another machine in same minicond env?
Thanks.

**Comment by chongxiaoc** [comment]
@niranjansuthar70 

You should have MPI and NFS installed successfully on your cluster. Then you only have to kick out the test with horovodrun -np n python train.py on head node.

example: https://medium.com/mpi-cluster-setup/mpi-clusters-within-a-lan-77168e0191b1

**Comment by niranjansuthar70** [comment]
Hello @chongxiaoc thanks for your response.  I have installed MPI and make a "storage" shared folder using NFS on 2 machines. I have one test.py files both machines and I have also modified hosts files with master and client ip addresses. Should I not add "`-H and master, client`" in command you sugggested" ?

**Comment by niranjansuthar70** [comment]
Hello @tgaddair , I have setup my project at same directory on both machines and I am able to run` "mpirun -np 1 -hosts client1 python train.py" `and ` "mpirun -np 1 -hosts master python train.py"` but when I run "mpirun -np 2 -hosts master:1, client15:1 python train.py" it gives me this error: 
`Host key verification failed.`

I am able to ssh from master to client and client to master passwordless.

**Comment by chongxiaoc** [comment]
> Hello @chongxiaoc thanks for your response. I have installed MPI and make a "storage" shared folder using NFS on 2 machines. I have one test.py files both machines and I have also modified hosts files with master and client ip addresses. Should I not add "`-H and master, client`" in command you sugggested" ?

Yep, because you are using NFS, `test.py` file is synced on all the machines, you only have to use `horovodrun` to execute it in parallel.

**Comment by niranjansuthar70** [comment]
Hello @chongxiaoc and @tgaddair I have resolved the previous issues and I am not using NFS instead I am making same directory on both machine and started distribution trianing on 2 machines and single training on 1 machine.
I am attaching the tensorboard analysis of both experiments.
My question is in image as you can see that DT_2 [distributiion on 2 machine] competed 2.7k steps in around 1 hr compare to single machine which takes around 29 mins to 2.7k steps. Is my distribution training on 2 machines working fine? or can you elaborate more on these steps and speed performance.?
![single machine training](https://user-images.githubusercontent.com/42465704/112261909-b2ee2f00-8c92-11eb-9bbe-043697b10392.png)


Thanks!
